[JIT] Add property support for ScriptModules

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42390 [JIT] Add property support for ScriptModules**

**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

Commit:
[JIT] Add property support for ScriptModules

**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

[ghstack-poisoned]

SplitInfinity commented on torch/jit/_recursive.py:
```diff
@@ -239,14 +252,6 @@ def infer_type(name, item):
                     "to a TorchScript type.)").format(torch.typename(type(value)))
             concrete_type_builder.add_failed_attribute(name, hint)
 
-    # Add @property methods as failed attributes, to give a better error message.
-    for name, value in type(nn_module).__dict__.items():
-        if isinstance(value, property):
-            hint = ("\n(This attribute exists on the Python module, but it's an @property "
-                    "method. @property methods are not yet supported in TorchScript. "
-                    "Please file a feature request on Github)")
-            concrete_type_builder.add_failed_attribute(name, hint)
```
I wasn't quite sure what to do here. Properties are really just syntactic sugar for setter and getter methods. Do they belong on the concrete type? If yes, how do we represent them? If not, what happens if two modules have the same underlying data attributes but different property getters/setters for them?

eellison replied:
So, concrete type already returns false if two classes differ: pyClass_.is(other.pyClass_). As long as we only take properties from methods, and not from attributes, we don't need to do any additional work here. If we do take properties as attributes than we'll need to add stuff to the ConcreteType. 

(i think it's fine to not take properties that are passed in as attributes, no one really does that)

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/python/python_tree_views.cpp:
```diff
@@ -162,7 +162,18 @@ void initTreeViewBindings(PyObject* module) {
+          return property.setter().get().name();
+        }
+
+        return Ident::create(property.range(), "");
```
Can we use None here instead ? empty string as null value could lead to some confusing debugging

SplitInfinity replied:
`None` as a string literal? I tried to use `py::none()` but that didn't work because of what I think is a problem inferring the return type of this lambda. Is there an example you can point me to?

eellison replied:
you can return None with c10::optional, and cast both returns to `c10::optional<Ident>` so that the return type gets properly inferred

SplitInfinity replied:
Fixed.

eellison commented on torch/jit/_recursive.py:
```diff
@@ -29,6 +31,16 @@
 
+ignored_properties = [
+    # Temporary fix for RNN module property named 'all_weights' being scripted
+    "all_weights",
```
Can we switch this instead to a list of (Python Class, [ignored_properties]) tuples, and we ignore a property if the scripted class is a subclass of our registered class ? right now we're ignoring these names indiscriminately

SplitInfinity replied:
Fixed. However, I had to use `torch.nn.Module` as the class for all of the `RecursiveScriptModule` properties that we need to ignore because of circular import problems.

eellison commented on torch/jit/_recursive.py:
```diff
@@ -407,14 +419,20 @@ def init_fn(script_module):
+    for property_stub in property_stubs:
+        property_name = property_stub.def_.name().name
+        fget = cpp_module._get_method(property_stub.def_.getter_name().name)
+        fset = cpp_module._get_method(property_stub.def_.setter_name().name)
```
don't we need to handle a setter not begin set ?

SplitInfinity replied:
Good catch, fixed.

eellison reviewed (commented):
Looks good, a few small comments

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

eellison commented on torch/jit/_recursive.py:
```diff
@@ -407,14 +424,21 @@ def init_fn(script_module):
+        fget = cpp_module._get_method(property_stub.def_.getter_name().name)
+        setter_name = property_stub.def_.setter_name()
+        fset = cpp_module._get_method(setter_name.name) if setter_name else None
+        script_module.__dict__[property_name] = property(property_name, fget, fset)
```
just for my own understanding, why was this needed ? maybe add a comment

eellison reviewed (approved):
(`-,-,
                                   ('(_,( )
                                    _   `_'
                                 __|_|__|_|_
                               _|___________|__
                              |o o o o o o o o/  
                             ~'`~'`~'`~'`~'`~'`~

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

eellison commented on torch/jit/_recursive.py:
```diff
@@ -548,6 +580,47 @@ def ignore_overloaded(method_name):
     return overload_stubs + stubs
 
+
+def should_compile_property(cls, property_name):
```
Now that we've added ignore, can we just add ignores on the modules themselves instead of maintaining a special list?

SplitInfinity replied:
Yeah good point.

SplitInfinity replied:
Sigh actually we can't do this because you can't dynamically add properties to an instance. You have to add them to the class. :'(

eellison commented on torch/jit/_recursive.py:
```diff
@@ -371,16 +371,22 @@ def init_fn(script_module):
+                unbound_setter = getattr(type(nn_module), item.fset.__name__) if item.fset else None
+                bound_setter = unbound_setter.__get__(script_module) if unbound_setter else None
+                setattr(script_moudle, name, property(name, bound_getter, bound_setter))
+            else:
+                pass
```
```
else:
     pass
```
is dead code

eellison commented on torch/jit/_recursive.py:
```diff
@@ -371,16 +371,22 @@ def init_fn(script_module):
                 bound_method = unbound_function.__get__(script_module)
                 setattr(script_module, name, bound_method)
+            elif isinstance(item, property) and _jit_internal.is_ignored_fn(item.fget):
+                unbound_getter = getattr(type(nn_module), item.fget.__name__)
```
why are we doing anything if `_jit_internal.is_ignored_fn(item.fget` ?

SplitInfinity replied:
Don't we still need to attach the property to the scripted module? Just as we do with ignored functions?

eellison replied:
yea, good point

eellison reviewed (commented):
a few questions

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -463,6 +463,15 @@ def forward(self, x):
 
+    if isinstance(drop, property):
+        prop = drop
+        setattr(prop.fget, "_torchscript_modifier", FunctionModifiers.IGNORE)
```
I had to use setattr due to mypy failures. A file for which mypy is enabled is typechecked, and mypy is instructed to recursive into this file to continue typechecking (I think). `# type : ignore` didn't work.

SplitInfinity replied:
The specific error in question was about setting an attr on a `Callable` that doesn't exist.

SplitInfinity replied:
Oh but using `setattr` causes a lint failure :(

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

SplitInfinity commented:
Ping @eellison 

I know you already approved this but the remaining items are more contentious than I thought so I just want to close them out.

eellison reviewed (approved):
LGTM

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add property support for ScriptModules"


**Summary**
This commit extends support for properties to include
ScriptModules.

**Test Plan**
This commit adds a unit test that has a ScriptModule with
a user-defined property.

`python test/test_jit_py3.py TestScriptPy3.test_module_properties`

Differential Revision: [D22880298](https://our.internmc.facebook.com/intern/diff/D22880298)

[ghstack-poisoned]
